timetable/extractor.py

```diff
@@ -0,0 +1,114 @@
+import json
+from dataclasses import dataclass
+from bs4 import BeautifulSoup
+from bs4.element import Tag, NavigableString, PageElement
+
+with open("full_location.json") as data_file:
+    full_locations = json.load(data_file)
+
+
+@dataclass
+class Course:
+    code: str
+    name: str
+    day: str
+    start_time: int
+    location: str
+    duration: int = 0
+
+    def get_location(self) -> str:
+        # TODO: more logic to specify other locations if possible
+        if ('NC' in self.location or 'NR' in self.location) and (
+                self.location[:2] + self.location[3]) in full_locations.keys():
+            return full_locations[self.location[:2] + self.location[3]]
+        if self.location in full_locations.keys():
+            return full_locations[self.location]
+        return self.location
+
+    @property
+    def title(self) -> str:
+        if 'NC' in self.location or 'NR' in self.location:
+            return f"[{self.location}] {self.name.title()}"
+        return self.name.title()
+
+    @property
+    def end_time(self) -> int:
+        return self.duration + self.start_time
+
+
+def build_courses(html: str, course_names: dict) -> list[Course]:
+    r"""
+    Build a list of Course objects from the html timetable
+    Args:
+        html: html code of the timetable returned by the server
+        course_names: mapping of course codes to course names
+
+    Returns:
+        list of Course objects
+    """
+    courses = []
+    days = {'Mon': "Monday", 'Tue': "Tuesday", 'Wed': "Wednesday", 'Thur': "Thursday", 'Fri': "Friday",
+            'Sat': "Saturday"}
+    soup = BeautifulSoup(html, 'html.parser')
+    table = soup.find('table', {'border': '1', 'cellpadding': '0', 'cellspacing': '0'})
+
+    def create_timings(_table: Tag | NavigableString) -> list[int]:
+        r"""# Creates a list of timings in 24 hours format - [8, ..., 12, 14, ..., 17]"""
+        headings = _table.find_all('td', {'class': 'tableheader',
+                                          'style': 'padding-top:5px;padding-bottom:5px;padding-left:7px;padding-right'
+                                                   ':7px',
+                                          'nowrap': ''})
```

Comment:
I don't think you have to specify all those properties of the `td`, only `{'class': 'tableheader'}` should suffice

Reply:
It will read the week days rows too in that case. I tried running the code and it is giving me an error. Please revert this change @Devansh-bit (commit karne se pehle test nahi kiya lol 👀)